editor/editor_file_system.cpp

```diff
@@ -139,6 +151,34 @@ Vector<String> EditorFileSystemDirectory::get_file_deps(int p_idx) const {
 	return deps;
 }
 
+String EditorFileSystemDirectory::get_file_resource_script_path(int p_idx) const {
+	ERR_FAIL_INDEX_V(p_idx, files.size(), "");
+	return files[p_idx]->resource_script_path;
+}
+
+String EditorFileSystemDirectory::get_file_resource_script_class_name(int p_idx) const {
+	ERR_FAIL_INDEX_V(p_idx, files.size(), "");
+	String script_path = files[p_idx]->resource_script_path;
+	String script_path_base_dir = script_path.get_base_dir();
+	String script_file = script_path.get_file();
+
+	if (script_path_base_dir == EditorFileSystemDirectory::get_path()) {
+		// If our resource script path happens to be in the same directory as our
+		// resource, then we can directly fetch the script's FileInfo.
+		int file_idx = EditorFileSystemDirectory::find_file_index(script_file);
+		ERR_FAIL_COND_V(file_idx < 0, "");
+		return files[file_idx]->script_class_name;
+	} else {
+		// If our resource script path is not in the same directory as our resource,
+		// we must fetch it's EditorFileSystemDirectory first, and then fetch the script's
+		// FileInfo from the newly fetched directory.
+		EditorFileSystemDirectory *dir = EditorFileSystem::get_singleton()->get_filesystem_path(script_path_base_dir);
+		int file_idx = dir->find_file_index(script_file);
+		ERR_FAIL_COND_V(file_idx < 0, "");
+		return dir->files[file_idx]->script_class_name;
+	}
```

Comment:
1. (unless you were just adding temporary notes for yourself) Generally prefer not including comments unless it specifically is providing additional contextual details for how it interacts with other code elsewhere or perhaps explains why it chose a specific algorithm over another one. The code should be written in such a way that it is largely self-documenting. Overuse of comments leads to not-maintained "documentation" embedded in the code that quickly becomes stale as code is modified over time in large projects.
2. Because `get_path()` and `find_file_index` are member functions (i.e. methods) of the `EditorFileSystemDirectory` class, and we are inside of that class currently, you do not need to preface them with the class name scope; it should automatically call the local implementation. You only do this for `EditorFileSystem::get_singleton()` because that specifically fetches a static instance of the `EditorFileSystem` class and thus can only be accessed via a static getter function to begin with.